### PR TITLE
Qingwei/refactor trade object

### PR DESCRIPTION
**Pull request by qingweibinary**

### CODE REVIEW

**Comment by qingweibinary** [comment]
@borisyankov do not merge until you read my slack pm ~
We need some other solution

**Comment by qingweibinary** [comment]
The last commit will fix the race condition mentioned https://trello.com/c/LgTpaPbx
The problem happens due to many small actions being triggered once Component is mounted, and we unmount the component before all actions are completed.

**Comment by qingweibinary** [comment]
@borisyankov this PR is ready, it breaks the chart after you purchase, but this is more correct, the updateExtremes call should not depends on trade and contract changes, as conceptually, Trade and Contract object does not change, at least from POV of chart, they should not change.

Meaning updateExtremes should be called on other places.

Problem ~

screen shot 2016-05-09 at 12 40 08 pm

The extremes is set wrongly (I believe when contract are 1st added into the chart) then it is never updated again since contract and trade values remain the same.

Let me know if you think it's this PR's mistake instead of binary-charts

**Comment by qingweibinary** [comment]
I forgot to test one part, I'll close it to prevent accidental merge

**Comment by qingweibinary** [comment]
Ok to be merge now

File: src/_reducers/TradesReducer.js [hunk 1]
```diff
@@ -9,18 +9,27 @@ import {
...
+        symbol: 'R_100',
+        tradeCategory: 'risefall',
+        duration: 5,
+        durationUnit: 't',
+        basis: 'stake',
+        amount: 50,
+        type: 'CALL',
+        barrierType: 'relative',
+    },
+    uiState: {
```
Comment (borisyankov) [inline src/_reducers/TradesReducer.js]: this does not sound like good name.

File: src/_reducers/TradesReducer.js [hunk 2]
```diff
@@ -9,18 +9,27 @@ import {
...
+        showAssetPicker: false,
```
Comment (borisyankov) [inline src/_reducers/TradesReducer.js]: This is now a local state for the picker and nothing to do with Reducers I think.

File: src/_reducers/TradesReducer.js (hunk 1 above)
Comment (qingweibinary) [inline src/_reducers/TradesReducer.js]: suggestion? things inside are like `showAssetPicker` which is totally ui state

File: src/fulltrade/FullTradeCard.js [hunk 3]
```diff
@@ -33,32 +36,34 @@ export default class FullTradeCard extends Component {
...
+        const { symbolName } = trade.params;
 
         const contract = this.props.contract || mockedContract;
 
         const contractAllowStartLater = findDeep(contract, child => child && !!child.forwardStartingDuration);
 
         const disabled =
             contract === mockedContract ||
-            trade.disabled ||
+            trade.uiState.disabled ||
```
Comment (borisyankov) [inline src/fulltrade/FullTradeCard.js]: I think you are moving in the wrong direction.
We need to make a better differentiation than a 'trade panel' and a trade object.
Trade object will be flat and has only stuff you need to perform the trade.
Trade panel can be what you are trying to do here.

File: src/_reducers/TradesReducer.js (hunk 2 above)
Comment (qingweibinary) [inline src/_reducers/TradesReducer.js]: Oh, did not notice, I'll verify and remove as needed

**Comment by qingweibinary** [comment]
@borisyankov 

I removed showAssetPicker state, I suggest we merge it 1st, and open another card for structural changes if needed, I think it would means splitting TradesReducer into 2 to 3 reducers, and wire them up in selectors, which should not affect most changes here, as the actions' interface will be the same.

**Pull request merged**